allennlp/data/dataset_readers/penn_tree_bank.py

```diff
@@ -0,0 +1,173 @@
+        are pre-tokenised in the data file.
+    """
+    def __init__(self,
+                 token_indexers: Dict[str, TokenIndexer] = None) -> None:
+        super().__init__(lazy=False)
+        self._token_indexers = token_indexers or {'tokens': SingleIdTokenIndexer()}
+
+    @overrides
+    def _read(self, file_path):
+        # if `file_path` is a URL, redirect to the cache
+        file_path = cached_path(file_path)
+
+        directory, filename = os.path.split(file_path)
+
+        logger.info("Reading instances from lines in file at: %s", file_path)
+        parses: List[Tree] = BracketParseCorpusReader(root=directory, fileids=[filename]).parsed_sents()
+
+        for parse in Tqdm.tqdm(parses):
+            yield self.text_to_instance(parse.leaves(), [x[1] for x in parse.pos()], parse)
+
+    @overrides
+    def text_to_instance(self, # type: ignore
+                         tokens: List[str],
+                         pos_tags: List[str],
+                         gold_tree: Tree = None) -> Instance:
+        """
+        We take `pre-tokenized` input here, because we don't have a tokenizer in this class.
+
+        Parameters
+        ----------
+        tokens : ``List[str]``, required.
+            The tokens in a given sentence.
+        pos_tags ``List[str]``, required.
+            The pos tags for the words in the sentence.
+        gold_tree : ``Tree``, optional (default = None).
+            The gold parse tree to create span labels from.
+
+        Returns
+        -------
+        An ``Instance`` containing the following fields:
+            tokens : ``TextField``
+                The tokens in the sentence.
+            pos_tags : ``SequenceLabelField``
+                The pos tags of the words in the sentence.
+            spans : ``ListField[SpanField]``
+                A ListField containing all possible subspans of the
+                sentence.
+            span_labels : ``SequenceLabelField``, optional.
+                The constiutency tags for each of the possible spans, with
+                respect to a gold parse tree. If a span is not contained
+                within the tree, a span will have a ``NO-LABEL`` label.
+        """
+        # pylint: disable=arguments-differ
+        text_field = TextField([Token(x) for x in tokens], token_indexers=self._token_indexers)
+        pos_tag_field = SequenceLabelField(pos_tags, text_field, "pos_tags")
+        fields = {"tokens": text_field, "pos_tags": pos_tag_field}
+        spans: List[Field] = []
+        gold_labels = []
+
+        if gold_tree is not None:
+            gold_spans_with_pos_tags = self._get_gold_spans(gold_tree, 0, OrderedDict()).items()
+            gold_spans = {span: label for (span, label)
+                          in gold_spans_with_pos_tags if "-POS" not in label}
+        else:
+            gold_spans = None
+        for start, end in enumerate_spans(tokens):
```

Comment:
This function takes some configuration parameters - is it worth allowing the `DatasetReader` to configure this?

Reply:
None of them really make sense here - we always want to consider all possible span sizes, because you can have single word non-pre-terminal spans, and we always want the full sentence span to be the root. The filter functions could be useful, but these can't be passed via the configuration file anyway, so i'll leave this as it is for now I think.